### PR TITLE
Raid Events

### PR DESCRIPTION
**Pull request by acikek**
Draft pull request. Current implementation functions, but is lacking event meta + an entry for the Raid Data map structure. It's also worth checking the organization of the additions before going further.

## Additions
- `player triggers raid` event
- `raid finishes` event
- `raid spawns wave` event
- `raid stops` event

All of these events have a `raid` context tag, and all of them except for the player event have support for the `in` switch for the raid location.

- `RaidData` utility class that turns data from `Raid` into a MapTag
  - The main question is: should all these events get one superclass event? It would negate the need for this utility class addition, and also provide better organization for the individual events (raid context would be handled in the superclass).

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerTriggersRaidScriptEvent.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    public boolean matches(ScriptPath path) {
+        if (!runInCheck(path, event.getPlayer().getLocation())) {
+            return false;
+        }
+        return super.matches(path);
+    }
+
+    @Override
+    public ObjectTag getContext(String name) {
+        if (name.equals("raid")) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/events/player/PlayerTriggersRaidScriptEvent.java]: use `switch` format not `if`

File: plugin/src/main/java/com/denizenscript/denizen/events/world/RaidStopsScriptEvent.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    public RaidStopsScriptEvent() {
+        registerCouldMatcher("raid stops");
+        registerSwitches("reason");
+    }
+
+    public RaidStopEvent event;
+
+    @Override
+    public boolean matches(ScriptPath path) {
+        if (path.switches.containsKey("reason") && !path.checkSwitch("reason", CoreUtilities.toLowerCase(event.getReason().name()))) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/events/world/RaidStopsScriptEvent.java]: use `runGenericSwitchCheck`, remove the containsKey check, remove the toLower - it automates all of this stuff

**Comment by acikek** [comment]
Changes since initial draft:
- Abandonded `RaidData` and created a generic `RaidScriptEvent` which provides `<context.raid>` and the meta entry for the MapTag format
- Filled out meta entries of each individual event

File: plugin/src/main/java/com/denizenscript/denizen/events/world/RaidScriptEvent.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    // @description
+    // Every event related to village raids has a <context.raid> property, a MapTag wrapper around raid data (see <@link url https://hub.spigotmc.org/javadocs/spigot/org/bukkit/Raid.html>).
+    // These events are <@link event player triggers raid>, <@link event raid finishes>, <@link event raid spawns wave>, and <@link event raid stops>.
+    //
+    // The data format is as follows:
+    // location: a LocationTag of the raid's center
+    // heroes: a list of PlayerTags that have participated in the raid
+    // raiders: a list of raider EntityTags that remain in the current wave
+    // status: the current status of the raid. See <@link url https://hub.spigotmc.org/javadocs/spigot/org/bukkit/Raid.RaidStatus.html>
+    // ticks: the raid's age in ticks
```
Comment (mergu) [inline plugin/src/main/java/com/denizenscript/denizen/events/world/RaidScriptEvent.java:32]: This should probably be turned into a duration

Comment (acikek) [reply plugin/src/main/java/com/denizenscript/denizen/events/world/RaidScriptEvent.java:32]: Updated

**Pull request merged**